patches/tModLoader/Terraria/Social/Steam/WorkshopHelper.TML.cs

```diff
@@ -279,77 +328,174 @@ internal class ModManager
+				foreach (var item in localMods) {
+					if (item.Enabled && GetPublishIdLocal(item, out ulong publishId))
+						list.Add(new PublishedFileId_t(publishId));
+				}
+				
+				int count = list.Count;
 				if (count == 0)
-					return true;
-				//TODO: Improve. You can't begin tracking more than 100 items within one call.
-				else if (count >= 100)
-					return false;
+					return;
+
+				int pg = count / PlaytimePagingConst;
+				int rem = count % PlaytimePagingConst;
+
+				for (int i = 0; i < pg + 1; i++) {
+					var pgList = list.GetRange(i * PlaytimePagingConst, (i == pg) ? rem : PlaytimePagingConst);
+
+					// Call the appropriate variant, may need performance optimization.
+					if (SteamUser)
+						SteamUGC.StartPlaytimeTracking(pgList.ToArray(), (uint)pgList.Count);
+					else
+						SteamGameServerUGC.StartPlaytimeTracking(pgList.ToArray(), (uint)pgList.Count);
+				}
+			}
 
+			public static void StopPlaytimeTracking() {
 				// Call the appropriate variant
 				if (SteamUser)
-					SteamUGC.StartPlaytimeTracking(modsById, (uint)modsById.Length);
+					SteamUGC.StopPlaytimeTrackingForAllItems();
 				else
-					SteamGameServerUGC.StartPlaytimeTracking(modsById, (uint)modsById.Length);
+					SteamGameServerUGC.StopPlaytimeTrackingForAllItems();
+			}
+		}
+
+		internal static class QueryHelper
+		{
+			internal const int QueryPagingConst = Steamworks.Constants.kNumUGCResultsPerPage;
+			internal static int IncompleteModCount;
+			internal static uint TotalItemsQueried;
+			internal static EResult ErrorState = EResult.k_EResultNone;
+
+			internal static List<ModDownloadItem> Items = new List<ModDownloadItem>();
+
+			internal static bool FetchDownloadItems() {
+				if (!QueryWorkshop())
+					return false;
 
 				return true;
 			}
 
-			public static bool StopPlaytimeTracking(PublishedFileId_t[] modsById) {
-				uint count = (uint)modsById.Length;
+			internal static ModDownloadItem FindModDownloadItem(string modName)
+			=> Items.FirstOrDefault(x => x.ModName.Equals(modName, StringComparison.OrdinalIgnoreCase));
 
-				if (count == 0)
-					return true;
-				//TODO: Improve. You can't begin tracking more than 100 items within one call.
-				else if (count >= 100)
+			internal static bool QueryWorkshop() {
+				IncompleteModCount = 0;
+				TotalItemsQueried = 0;
+				Items.Clear();
+				ErrorState = EResult.k_EResultNone;
+
+				AQueryInstance.InstalledMods = ModOrganizer.FindMods();
+
+				if (!ModManager.SteamAvailable)
 					return false;
 
-				// Call the appropriate variant
-				if (SteamUser)
-					SteamUGC.StopPlaytimeTracking(modsById, (uint)modsById.Length);
-				else
-					SteamGameServerUGC.StopPlaytimeTracking(modsById, (uint)modsById.Length);
+				if (!new AQueryInstance().GetPageCountFast())
+					return false;
+
+				//TODO: Threaded query of multiple pages at a time appears un-supported within Steam.
+				// Concerned for future speed of querying the Steam Workshop (ie at 1000+ items).
```

Comment:
Have you tried using the version which uses a cursor[1][2] instead of page number? It _may_ be faster[3], and I think the current method might be limited to 1000 pages total[3]. Should also cut out some pagination logic.

[1] https://store.steampowered.com/news/group/4145017/old_view/1751242076600910671
[2] https://github.com/rlabrecque/Steamworks.NET/blob/9416761cab7fa6b226803c95a0cc7ab8dd03b43c/CodeGen/steam/isteamugc.h#L211
[3] https://github.com/rlabrecque/Steamworks.NET/blob/9416761cab7fa6b226803c95a0cc7ab8dd03b43c/CodeGen/steam/isteamugc.h#L386

Reply:
If this turns out to be parallelisable, I imagine that would be a lot faster. But if not this is probably worth pursuing.

Reply:
Hmm, I don't see that in the standard documentation: https://partner.steamgames.com/doc/api/ISteamUGC#CreateQueryAllUGCRequest
That said, when the release notes say its deprecated, and with the upgrade Mirsario did to Steamworks.Net 15, I guess that is the new proper method.
Thanks for pointing this out, much easier to deal with now than later.

Reply:
Made the switch, improved the speed somewhat, but if the method is deprecated and subject to removal, its best we don't keep using the old method. 

Current performance is roughly 0.4 seconds per 50 items (or 2 seconds per 250 items). Not bad, but at 2000 items plus, that's 16 seconds of waiting. Still a concern, and parallellization still isn't possible.